### PR TITLE
[cudamapper] Improved PAF output

### LINKED ISSUE #410: [cudamapper] Improve PAF generation
New implementation of index caching (#318) is bottlenecked by generation of strings for PAF files (not writing to files themselves). We currently copy read names and lengths into `Overlap` objects using `Overlapper::update_read_names()` and then use `Overlapper::print_paf()`.

Remove read name and length from `Overlap` object and parallelize `Overlapper::print_paf()` for better performance.

### PR DESCRIPTION
**Pull request by mimaric**
This PR changes `Overlap` structure so it does not contain read names anymore. They are now being directly read from `FastaParser` based on `Overlap`'s `read_id`.
This means `Overlap::clear()` and `update_read_names()` are not needed any more.

This change also brings CPU multithreading to `print_paf()`.

As we discussed will will have to further improve `print_paf()`'s performance, but filing this PR separately as it it brings some improvements already and cleans up `Overlap`.

Closes #410 

@edawson I don't think this intersect with your work, but FYI

@vellamike Can you please make sure I did not mess up something with your PR #415

### CODE REVIEW

File: cudamapper/tests/Test_CudamapperOverlapperTriggered.cu [hunk 1]
```diff
@@ -11,144 +11,12 @@
...
-#include <random>
-
-#include "cudamapper_file_location.hpp"
 
 namespace claragenomics
 {
 namespace cudamapper
 {
 
-TEST(TestCudamapperOverlapperTriggerred, FuseTwoOverlaps)
```
Comment (ahehn-nv) [inline cudamapper/tests/Test_CudamapperOverlapperTriggered.cu:29]: Why were these tests removed? They seem unrelated to the changes in the main code. Do we have better tests for this functionality somewhere else?

Comment (mimaric) [reply cudamapper/tests/Test_CudamapperOverlapperTriggered.cu:29]: These tests only test `fused_overlaps()` function from `cudamapper_utils.hpp`. This function became unused at some point, probably when @vellamike introduced `Overlapper::post_process_overlaps()` so I also removed the function from `cudamapper_utils.hpp`.
I agree this could have been a separate PR as well.

File: cudamapper/src/cudamapper_utils.cpp [hunk 2]
```diff
@@ -8,62 +8,97 @@
...
-            auto tmp                         = o.target_end_position_in_read_;
-            o.target_end_position_in_read_   = o.target_start_position_in_read_;
-            o.target_start_position_in_read_ = tmp;
-        }
-        else
-        {
-            o.relative_strand = RelativeStrand::Forward;
-        }
-    };
+    assert(!cigar.empty() || (overlaps.size() && cigar.size()));
```
[the remark this answers is not in the document]
Comment (mimaric) [reply cudamapper/src/cudamapper_utils.cpp:31]: Whoops, yes...
done

File: cudamapper/src/cudamapper_utils.cpp [hunk 3]
```diff
@@ -8,62 +8,97 @@
...
+                                                    overlaps[i].num_residues_ * kmer_size, // Print out the number of residue matches multiplied by kmer size to get approximate number of matching bases
+                                                    std::max(std::abs(static_cast<std::int64_t>(overlaps[i].target_start_position_in_read_) - static_cast<std::int64_t>(overlaps[i].target_end_position_in_read_)),
+                                                             std::abs(static_cast<std::int64_t>(overlaps[i].query_start_position_in_read_) - static_cast<std::int64_t>(overlaps[i].query_end_position_in_read_))), //Approximate alignment length
+                                                    255);
+            chars_in_buffer += added_chars;
+
+            // If CIGAR string is generated, output in PAF.
+            if (!cigar.empty())
+            {
+                added_chars = std::sprintf(buffer.data() + chars_in_buffer,
```
[the remark this answers is not in the document]
Comment (mimaric) [reply cudamapper/src/cudamapper_utils.cpp:86]: Agree, done

File: cudamapper/src/cudamapper_utils.cpp [hunk 4]
```diff
@@ -8,62 +8,97 @@
...
-        {
-            set_relative_strand(fused_overlap);
-            fused_overlaps.push_back(fused_overlap);
-            fused_overlap = unfused_overlaps[i + 1];
+            const std::string& query_read_name  = query_parser.get_sequence_by_id(overlaps[i].query_read_id_).name;
+            const std::string& target_read_name = target_parser.get_sequence_by_id(overlaps[i].target_read_id_).name;
+
+            // (over)estimate the number of character that are going to be needed
+            // 150 is an overestimate of number of characters that are going to be needed for non-string values
+            std::int32_t expected_chars = 150 + query_read_name.length() + target_read_name.length();
```
[the remark this answers is not in the document]
Comment (mimaric) [reply cudamapper/src/cudamapper_utils.cpp:53]: done

File: cudamapper/src/cudamapper_utils.cpp [hunk 5]
```diff
@@ -8,62 +8,97 @@
...
+            if (!cigar.empty())
+            {
+                expected_chars += cigar[i].length();
```
[the remark this answers is not in the document]
Comment (mimaric) [reply cudamapper/src/cudamapper_utils.cpp:56]: done

File: cudamapper/src/cudamapper_utils.cpp [hunk 6]
```diff
@@ -8,62 +8,97 @@
...
+            }
+
+            // if there is not enough space in buffer reallocate
+            if (buffer.size() - chars_in_buffer < expected_chars)
```
[the remark this answers is not in the document]
Comment (mimaric) [reply cudamapper/src/cudamapper_utils.cpp:60]: done

File: cudamapper/src/cudamapper_utils.cpp [hunk 7]
```diff
@@ -8,62 +8,97 @@
...
+    assert(!cigar.empty() || (overlaps.size() && cigar.size()));
 
-    if (unfused_overlaps.size() == 0)
+#pragma omp parallel
     {
-        return;
-    }
-
-    Overlap fused_overlap = unfused_overlaps[0];
+        std::size_t number_of_threads = omp_get_num_threads();
```
[the remark this answers is not in the document]
Comment (mimaric) [reply cudamapper/src/cudamapper_utils.cpp:35]: done

File: cudamapper/src/cudamapper_utils.cpp [hunk 8]
```diff
@@ -8,62 +8,97 @@
...
+            // Add new line to demarcate new entry.
+            buffer.data()[chars_in_buffer] = '\n';
+            ++chars_in_buffer;
         }
-    }
+        buffer[chars_in_buffer] = '\0';
 
-    set_relative_strand(fused_overlap);
-    fused_overlaps.push_back(fused_overlap);
+        std::lock_guard<std::mutex> lg(write_output_mutex);
```
[the remark this answers is not in the document]
Comment (mimaric) [reply cudamapper/src/cudamapper_utils.cpp:98]: Yes, I'll look into using `std::thread`/`std::async` here

File: cudamapper/tests/Test_CudamapperOverlapperTriggered.cu (hunk 1 above)
Comment (ahehn-nv) [reply cudamapper/tests/Test_CudamapperOverlapperTriggered.cu:29]: No need for a separate PR - it was just a bit confusing.

File: cudamapper/src/cudamapper_utils.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mimaric) [reply cudamapper/src/cudamapper_utils.cpp]: Done using `std::thread`. Note that string formatting will likely eventually be moved to GPU.

[the remark this answers is not in the document]
Comment (ahehn-nv) [reply cudamapper/src/cudamapper_utils.cpp]: good point. Didn't consider this while I was reviewing...

File: cudamapper/src/cudamapper_utils.cpp [hunk 9]
```diff
@@ -8,62 +8,143 @@
...
+        // If CIGAR string is generated, output in PAF.
+        if (!cigar.empty())
+        {
+            const int32_t added_cigar_chars = std::sprintf(buffer.data() + chars_in_buffer,
+                                                           "\tcg:Z:%s",
+                                                           cigar[i].c_str());
+            chars_in_buffer += added_cigar_chars;
+        }
+        // Add new line to demarcate new entry.
+        buffer.data()[chars_in_buffer] = '\n';
```
Comment (ahehn-nv) [inline cudamapper/src/cudamapper_utils.cpp:102]: Optional: While correct `buffer.data()[chars_in_buffer]` instead of `buffer[chars_in_buffer]` looks funny. Probably not intentional...

File: cudamapper/src/cudamapper_utils.cpp [hunk 10]
```diff
@@ -8,62 +8,143 @@
...
+    assert(!cigar.empty() || (overlaps.size() == cigar.size()));
+
+    // divide the work into several threads
+
+    int32_t number_of_threads   = std::thread::hardware_concurrency() / number_of_devices; // We could use a better heuristic here
+    int64_t overlaps_per_thread = get_size<int64_t>(overlaps) / number_of_threads;
+
+    std::vector<std::thread> threads;
+
+    for (uint32_t thread_id = 0; thread_id < number_of_threads; ++thread_id)
```
Comment (ahehn-nv) [inline cudamapper/src/cudamapper_utils.cpp:130]: Optional: Why `uint32_t` here all of the sudden?

File: cudamapper/src/cudamapper_utils.cpp (hunk 9 above)
Comment (mimaric) [reply cudamapper/src/cudamapper_utils.cpp:102]: Yup, probably remained from the time I was playing around with it.
Done

File: cudamapper/src/cudamapper_utils.cpp (hunk 10 above)
Comment (mimaric) [reply cudamapper/src/cudamapper_utils.cpp:130]: Again, initially I was using `uint32_t`, then I switched to `int32_t` and forgot to change
Done
